lib/crash_reporter.js

```diff
@@ -0,0 +1,191 @@
+          }
+        });
+      })
+      .catch(function(error) {
+        reject(error);
+      });
+  });
+  return promise;
+};
+
+Preferences.prototype.readPreferences = function() {
+  var promise = new Promise(function(resolve, reject) {
+    fs.exists(preferenceFile, function(exists) {
+      if (exists) {
+        fs.readFile(preferenceFile, function(error, data) {
+          if (error) {
+            reject(error);
+          } else {
+            resolve(JSON.parse(data));
+          }
+        });
+      } else {
+        // we don't have any local preferences
+        // return falsy value
+        resolve();
+      }
+    });
+  });
+  return promise;
+};
+
+// the only instance of the crash reporter
+var crashReporter = null;
+
+var CrashReporter = function() {
+  this.preferences = new Preferences();
+};
+
+CrashReporter.getInstance = function() {
+  if (crashReporter) {
+    return crashReporter;
+  } else {
+    crashReporter = new CrashReporter();
+    return crashReporter;
+  }
+};
+
+CrashReporter.prototype.turnOff = function() {
+  return this.preferences
+    .write(CRASH_REPORTER_PREFERENCE, 'off')
+    .catch(function(error) {
+      // do nothing
+      // do not crash the crash reporter :)
+      logs.err('Error turning off crash reporter preferences', error);
+    });
+};
+
+CrashReporter.prototype.turnOn = function() {
+  return this.preferences
+    .write(CRASH_REPORTER_PREFERENCE, 'on')
+    .catch(function(error) {
+      // do nothing
+      // do not crash the crash reporter :)
+      logs.err('Error turning on crash reporter preferences', error);
+    });
+};
+
+CrashReporter.prototype.submit = function(report) {
+  var that = this;
+  that.preferences.read(CRASH_REPORTER_PREFERENCE, 'on')
+    .then(function(value) {
+      if (value === 'on') {
+        var labels = [packageJson.name, packageJson.version].join(',');
+        that.submitReport(labels, report)
+          .then(function() {
+            logs.info('Done !');
+          })
+          .catch(function(error) {
+            logs.err('Error submitting crash report', error);
+          });
+      }
+    })
+    .catch(function(error) {
+      // do nothing
+      // do not crash the crash reporter :)
+      logs.err('Error submitting crash report', error);
+    });
+};
+
+CrashReporter.prototype.submitReport = function(labels, report) {
+  var promise = new Promise(function(resolve, reject) {
+    request.post({
+      'url': SUBMIT_CRASH_URL,
+      'form': {
+        'crash': report,
+        'labels': labels,
+        'f': 'json'
+      }
+    }, function(error, httpResponse, body) {
+      try {
+        if (error) {
+          reject(error);
+        } else {
+          var json = JSON.parse(body);
+          if (json['error']) {
+            reject(json['error']);
+          } else {
+            resolve();
+          }
+        }
+      } catch (exception) {
+        reject(exception);
+      }
+    });
+  });
+  return promise;
```

Comment:
In the rest of the code, we usually follow the convention of `return new Promise(...` rather than creating a local `promise` variable and immediately returning it.